### PR TITLE
Issue #1451 Logging general object

### PR DESCRIPTION
**Pull request by aarb1**
Issue #1451 Created a generic data object to replace specific data objects for log events.

### CODE REVIEW

**Comment by aarb1** [comment]
#build

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils.java [hunk 1]
```diff
@@ -12,16 +12,14 @@
 
 import java.io.UnsupportedEncodingException;
 import java.net.URLDecoder;
-import java.security.AccessController;
-import java.security.PrivilegedAction;
-import java.text.SimpleDateFormat;
+import java.util.ArrayList;
 
 import com.ibm.websphere.ras.DataFormatHelper;
 import com.ibm.ws.health.center.data.HCGCData;
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils.java]: need to get rid of HCGCData - should use GenericData for that as well.

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+        for (Pair p : pairs) {
+
+            if (p instanceof KeyValuePair) {
+
+                kvp = (KeyValuePair) p;
+                key = kvp.getKey();
+                value = kvp.getValue();
+
+                if (!(key.equals(LogFieldConstants.SEVERITY))) {
+
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java]: would be easier to understand method overall if you did this...
if (key.equals(LogFieldConstants.SEVERITY))) {}
else if (key.equals(...

**Comment by aarb1** [comment]
#build

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+import com.ibm.websphere.ras.DataFormatHelper;
+import com.ibm.ws.health.center.data.HCGCData;
+import com.ibm.ws.logging.data.GenericData;
+import com.ibm.ws.logging.data.KeyValuePair;
+import com.ibm.ws.logging.data.Pair;
+
+/**
+ *
+ */
+public class CollectorJsonUtils1_1 {
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java]: please create follow on issue to make methods type-safe below (change Object event to GenericData event and remove explicit casts to GenericData) once everything is using GenericData events.  Also do for CollectorJsonUtils.

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+            if (p instanceof KeyValuePair) {
+
+                kvp = (KeyValuePair) p;
+                key = kvp.getKey();
+                value = kvp.getValue();
+
+                if (key.equals(LogFieldConstants.SEVERITY)) {
+                }
+
+                if (key.equals(LogFieldConstants.MESSAGE)) {
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java]: this should be an else if

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/data/KeyValuePair.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    private final ValueTypes valueType;
+
+    public KeyValuePair(String key, String value, ValueTypes valueType) {
+        this.key = key;
+        this.value = value;
+        this.valueType = valueType;
+    }
+
+    public boolean isNumber() {
+        if (valueType == ValueTypes.NUMBER) {
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/data/KeyValuePair.java]: simplify... return (valueType == ValueTypes.NUMBER)

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/data/KeyValuePair.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    public boolean isString() {
+        if (valueType == ValueTypes.STRING) {
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/data/KeyValuePair.java]: simplify... return (valueType == ValueTypes.STRING)

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/source/LogSource.java [hunk 7]
```diff
@@ -109,28 +122,66 @@ public void publish(RoutedMessage routedMessage) {
...
+
+    public GenericData parse(RoutedMessage routedMessage) {
+
+        GenericData genData = new GenericData();
+        LogRecord logRecord = routedMessage.getLogRecord();
+        String messageVal = extractMessage(routedMessage, logRecord);
+
+        long dateVal = logRecord.getMillis();
+        // Must pass datetime as string, as its value type must be set as string in its kvp
+        genData.addPair("ibm_datetime", Long.toString(dateVal));
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/source/LogSource.java]: just add ibm_datetime as a long dateVal... it's formatter's job to decide if we want it to be a formatted like a string.

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/source/TraceSource.java [hunk 8]
```diff
@@ -86,33 +88,48 @@ public String getLocation() {
...
+        String verboseMessage = routedMessage.getFormattedVerboseMsg();
+        if (verboseMessage == null) {
+            genData.addPair("message", logRecord.getMessage());
+        } else {
+            genData.addPair("message", verboseMessage);
+        }
+
+        long datetimeValue = logRecord.getMillis();
+        // Must pass datetime as string, as its value type must be set as string in its kvp
+        genData.addPair("ibm_datetime", Long.toString(datetimeValue));
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/source/TraceSource.java]: just add ibm_datetime as a long dateVal... it's formatter's job to decide if we want it to be a formatted like a string.

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+                    } else if (key.equals(LogFieldConstants.IBM_THREADID)) {
+
+                        isFirstField = isFirstField
+                                       & !CollectorJsonHelpers.addToJSON(sb, key, DataFormatHelper.padHexString(Integer.parseInt(value), 8), false, true, false, isFirstField,
+                                                                         kvp.isNumber());
+
+                    } else if (key.equals(LogFieldConstants.IBM_DATETIME)) {
+
+                        String datetime = CollectorJsonHelpers.dateFormatTL.get().format(Long.parseLong(value));
+                        isFirstField = isFirstField & !CollectorJsonHelpers.addToJSON(sb, key, datetime, false, true, false, isFirstField, kvp.isNumber());
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java]: change last parm to insist it be treated as a String

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+                    if (userAgent != null && userAgent.length() > MAX_USER_AGENT_LENGTH) {
+                        userAgent = userAgent.substring(0, MAX_USER_AGENT_LENGTH);
+                    }
+
+                    isFirstField = isFirstField & !CollectorJsonHelpers.addToJSON(sb, key, userAgent, false, false, false, isFirstField, kvp.isNumber());
+
+                } else if (key.equals(LogFieldConstants.IBM_DATETIME)) {
+
+                    String datetime = CollectorJsonHelpers.dateFormatTL.get().format(Long.parseLong(value));
+                    isFirstField = isFirstField & !CollectorJsonHelpers.addToJSON(sb, key, datetime, false, true, false, isFirstField, kvp.isNumber());
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java]: change last parm to insist it be treated as a String

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+                } else if (key.equals(LogFieldConstants.IBM_THREADID)) {
+
+                    isFirstField = isFirstField
+                                   & !CollectorJsonHelpers.addToJSON(sb, key, DataFormatHelper.padHexString(Integer.parseInt(value), 8), false, true, false, isFirstField,
+                                                                     kvp.isNumber());
+
+                } else if (key.equals(LogFieldConstants.IBM_DATETIME)) {
+
+                    String datetime = CollectorJsonHelpers.dateFormatTL.get().format(Long.parseLong(value));
+                    isFirstField = isFirstField & !CollectorJsonHelpers.addToJSON(sb, key, datetime, false, true, false, isFirstField, kvp.isNumber());
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java]: change last parm to insist it be treated as a String

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/logging/source/AccessLogSource.java [hunk 12]
```diff
@@ -109,31 +109,36 @@ public void process(AccessLogRecordData recordData) {
...
+                genData.addPair("ibm_requestHost", recordData.getLocalIP());
+                genData.addPair("ibm_requestPort", recordData.getLocalPort());
+                genData.addPair("ibm_remoteHost", recordData.getRemoteAddress());
+                genData.addPair("ibm_userAgent", request.getHeader("User-Agent").asString());
+                genData.addPair("ibm_requestProtocol", request.getVersion());
+                genData.addPair("ibm_bytesReceived", recordData.getBytesWritten());
+                genData.addPair("ibm_responseCode", response.getStatusCodeAsInt());
+                genData.addPair("ibm_elapsedTime", recordData.getElapsedTime());
+                // Must pass datetime as string, as its value type must be set as string in its kvp
+                genData.addPair("ibm_datetime", Long.toString(recordData.getTimestamp()));
```
Comment (donbourne) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/logging/source/AccessLogSource.java]: just add ibm_datetime as a long dateVal... it's formatter's job to decide if we want it to be a formatted like a string.

**Comment by aarb1** [comment]
#build

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils.java [hunk 13]
```diff
@@ -139,7 +139,9 @@ private static String jsonifyGCEvent(int maxFieldLength, String wlpUserDir,
                 key = kvp.getKey();
                 value = kvp.getValue();
 
-                if (key.equals(LogFieldConstants.IBM_DURATION)) {
+                if (key.equals(LogFieldConstants.IBM_SEQUENCE)) {
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils.java]: not clear why IBM_SEQUENCE should be omitted.

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java [hunk 14]
```diff
@@ -130,7 +129,8 @@ private static String jsonifyGCEvent(int maxFieldLength, String wlpUserDir,
                 key = kvp.getKey();
                 value = kvp.getValue();
 
-                if (key.equals(LogFieldConstants.IBM_DURATION)) {
+                if (key.equals(LogFieldConstants.IBM_SEQUENCE)) {
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java]: not clear why IBM_SEQUENCE should be omitted.

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils.java [hunk 15]
```diff
@@ -93,367 +97,273 @@ public static String jsonifyGCEvent(String hostName, String wlpUserDir, String s
...
-                                    isFirstField);
-        isFirstField = isFirstField & !addToJSON(sb, isHigherVer ? "ibm_messageId" : "messageId", messageLogData.getMessageID(), false, false, false, isFirstField);
-        isFirstField = isFirstField & !addToJSON(sb, isHigherVer ? "module" : "loggerName", messageLogData.getLoggerName(), false, false, false, isFirstField);
-        isFirstField = isFirstField & !addToJSON(sb, isHigherVer ? "ibm_methodName" : "methodName", messageLogData.getMethodName(), false, false, false, isFirstField);
-        isFirstField = isFirstField
-                       & !addToJSON(sb, isHigherVer ? "ibm_threadId" : "threadId", DataFormatHelper.padHexString(messageLogData.getThreadID(), 8), false, false, false,
-                                    isFirstField);
-        isFirstField = isFirstField & !addToJSON(sb, isHigherVer ? "ibm_className" : "className", messageLogData.getClassName(), false, false, false, isFirstField);
-        isFirstField = isFirstField & !addToJSON(sb, "message", message, false, true, false, isFirstField);
+                if (key.equals(LogFieldConstants.IBM_SEQUENCE)) {
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils.java]: not clear why IBM_SEQUENCE should be omitted.

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+
+        for (Pair p : pairs) {
+
+            if (p instanceof KeyValuePair) {
+
+                kvp = (KeyValuePair) p;
+                key = kvp.getKey();
+                value = kvp.getValue();
+
+                if (key.equals(LogFieldConstants.IBM_SEQUENCE)) {
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java]: not clear why IBM_SEQUENCE should be omitted.

**Comment by aarb1** [comment]
#build

**Comment by aarb1** [comment]
#build

**Pull request closed without merging**